File: Math & Geometry/2083. Substrings That Begin and End With the Same Letter/2083.cpp
#pragma GCC optimize("O3","Oz","unroll-loops","inline-functions","fast-math","march=native","lto","omit-frame-pointer")
class Solution {
public:
    long long numberOfSubstrings(string s) {
        ios_base::sync_with_stdio(0);
        cin.tie(0);
        cout.tie(0);

        int n = s.size();
        long long answer = 0;
        vector<long long> prefixCount(26, 0);

        for (int i = 0; i < s.size(); i++) {
            prefixCount[s[i] - 'a']++;
            answer += prefixCount[s[i] - 'a'];
        }

        return answer;
    }
};

File: Math & Geometry/2083. Substrings That Begin and End With the Same Letter/2083.py
class Solution:
    def numberOfSubstrings(self, s: str) -> int:
        n = len(s)
        answer = 0
        prefix_count = [0] * 26

        for i in range(len(s)):
            prefix_count[ord(s[i]) - ord("a")] += 1
            answer += prefix_count[ord(s[i]) - ord("a")]

        return answer